src/lazy/expanded/macro_evaluator.rs

```diff
@@ -316,15 +317,54 @@ pub enum ValueExpr<'top, D: Decoder> {
 }
 
 impl<'top, D: Decoder> ValueExpr<'top, D> {
-    /// If this `ValueExpr` represents an entity encoded in te data stream, returns `Some(range)`.
+    pub fn expect_value_literal(&self) -> IonResult<LazyExpandedValue<'top, D>> {
+        match self {
+            ValueExpr::ValueLiteral(value) => Ok(*value),
+            _ => {
+                IonResult::decoding_error("expected a value literal, but found a macro invocation")
+            }
+        }
+    }
+
+    pub fn expect_macro_invocation(&self) -> IonResult<MacroExpr<'top, D>> {
+        match self {
+            ValueExpr::MacroInvocation(invocation) => Ok(*invocation),
+            _ => {
+                IonResult::decoding_error("expected a macro invocation, but found a value literal")
+            }
+        }
+    }
+
+    /// Returns `true` if this expression was produced by evaluating a macro. Otherwise, returns `false`.
+    pub fn is_ephemeral(&self) -> bool {
```

Comment:
I've been using it in `ion-rust` and `ion inspect` to describe values that don't appear anywhere in the input stream. They exist only during macro evaluation. I'm open to other ideas.

Reply:
Can you expand on "don't appear anywhere in the input stream"? An example, perhaps?

Reply:
Sure!

Here's an example in which the stream contains:
1. An encoding directive (`$ion_encoding::(...)`) that defines a macro called `triple`
2. An e-expression with a string argument (`(:triple "hello")`)

![image](https://github.com/user-attachments/assets/8dff3ac4-ec75-413d-ac81-8ace65e92aab)


When a `Reader` traverses the stream it will hand out three `LazyValue`s, one for each `"hello"` in the stream's expansion. However, if you were to ask any of those `LazyValue`s for its `Span` (backing bytes), you would get `None`--the values aren't encoded as values the stream itself, they're the ephemeral result of evaluating the e-expression.

Does that help?

Reply:
Thanks for the example :) Yes that does help. If the `triple` macro produced only one value then you could say that that is encoded where the macro is invoked (as you describe in the comment below about "singleton macros". 

Hmm, I could be inclined to think of those `"hello"` values as occupying the same 0-byte span between bytes... 73 and 74? But then I'd have to wonder- how do I distinguish the position of the second from the first? If I'm able to distinguish the difference between multiple values in the same _place_ then I might as well say all three values are "in" the macro invocation. Is the answer then that we need another dimension to locate values, that all three `"hello"` are located between bytes 67-74 at different ordinals? E.g. `([67-74], 0)`, `([67-74], 1)`, `([67-74], 2)`?

Are these values ephemeral? Virtual? Implicit? Derived? Computed?

Lots of questions. No suggestions :) Works for me.